### PR TITLE
CRM-20421: Inherited memberships are converted to individual memberships when the "parent" for the membership is merged with another contact

### PR DESCRIPTION
**Pull request by agilewarealok**
Overview
----------------------------------------
Inherited memberships are converted to individual memberships when the "parent" for the membership is merged with another contact.

**Steps to reproduce:**

- Create an organisation based Membership Type
- Set Membership Type inheritance based on "Employer of" relationship
- Create Organisation A
- Create Individual A
- Create relationship "Employer of" from Organisation A to Individual A
- Create Membership for Organisation A
- Note, inherited Membership automatically created for Individual A
- Create Organisation B
- Merge Organisation A and Organisation B

**Bug:** Individual A now has individual Membership which was not inherited by Parent
**Expected result:** Membership should have been removed.

Before
----------------------------------------
Inherited Memberships are not deleted when merging the contacts without copying the memberships to the other contact.

After
----------------------------------------
It now works as expected and inherited memberships get removed while merging the contact if we chose not to transfer the memberships.

_Agileware Ref: CIVICRM-552_

---

 * [CRM-20421: CIVICRM-552 Inherited memberships are converted to individual memberships when the "parent" for the membership is merged with another contact](https://issues.civicrm.org/jira/browse/CRM-20421)

### CODE REVIEW

**Comment by KarinG** [comment]
@davejenx ^^

**Comment by monishdeb** [comment]
@agilewarealok  I have tested your patch and fixes the issue by deleting the individual membership. However, on code pov I have prepared a [patch](https://gist.github.com/monishdeb/671009aa927120703bb8225c169c285d) on top of your changes, where I have made the following changes:
1. Used caps for id i.e. $membershipID instead of $membershipId
2. Optimised code by using chain API. 
3. Used CRM_Utils_Array::value('id', $values) this will collect 'id' values from array i.e.
```
for array('id' => 1, 'id' => 3, ..) return array(1,3,...)
```
3. Removed parameter $mainId as it is of no use
4. Some additional changes

Instead of that made few comments on added unit test.

File: tests/phpunit/api/v3/ContactTest.php [hunk 1]
```diff
@@ -1493,6 +1493,76 @@ public function testUpdateOrganizationWithAll() {
   }
 
   /**
+   * Test merging 2 organizations.
+   *
+   * This test make sure that inherited memberships are deleted upon merging organization.
+   */
+  public function testMergeOrganizations() {
+    $organizationid = $this->organizationCreate(array(), 0);
+    $organization2id = $this->organizationCreate(array(), 1);
```
Comment (monishdeb) [inline tests/phpunit/api/v3/ContactTest.php]: Use variable name $organizationID1 and $organizationID2 instead

File: tests/phpunit/api/v3/ContactTest.php [hunk 2]
```diff
@@ -1493,6 +1493,76 @@ public function testUpdateOrganizationWithAll() {
   }
 
   /**
+   * Test merging 2 organizations.
+   *
+   * This test make sure that inherited memberships are deleted upon merging organization.
```
Comment (monishdeb) [inline tests/phpunit/api/v3/ContactTest.php]: Mention the JIRA ticket as ```CRM-20421: This test make ..```

**Comment by jusfreeman** [comment]
@monishdeb thanks for reviewing!

**Comment by agilewarealok** [comment]
@monishdeb We've updated the PR with mentioned changes.

**Comment by jusfreeman** [comment]
@eileenmcnaughton or @seamuslee001 would you mind reviewing this one?

**Comment by colemanw** [comment]
* ([`r-jira`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-jira)) __Pass__
* ([`r-test`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-test)) __Pass__
* ([`r-code`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-code)) __Pass__
* ([`r-doc`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-doc)) __Pass__
* ([`r-maint`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-maint)) __Pass__
* ([`r-run`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-run)) __Pass__: Run by Monish
* ([`r-user`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-user)) __Pass__: Works as expeced
* ([`r-tech`](https://docs.civicrm.org/dev/en/latest/standards/review/#r-tech)) __Pass__: Includes a test :)

**Pull request merged**